### PR TITLE
Add Undo and Redo commands

### LINKED ISSUE #19: As a social worker, I can undo and redo my commands
- Add History, State and HistoryManager classes to update and keep track of current state
- History should be cleared upon exit and initialized upon start of application

### PR DESCRIPTION
**Pull request by pjanthony2001**
Added initial classes required to implement the Undo and Redo commands. Fixes #19

### CODE REVIEW

File: src/main/java/seedu/address/model/ModelManager.java [hunk 1]
```diff
@@ -121,12 +147,100 @@ public void setPerson(Person target, Person updatedPerson) {
...
+     * @throws HistoryException if error while making deep copy
+     */
+    public State generateState(Command command) throws HistoryException {
+        try {
+            return new State(command,
+                    getAddressBook().deepCopy(),
+                    deepCopyFilteredPersonsListSource(),
+                    getFilteredPersonsListPredicate());
+        } catch (IllegalValueException e) {
+            throw new HistoryException("Error while creating deepcopy", e);
```
Comment (Jolonauh) [inline src/main/java/seedu/address/model/ModelManager.java:214]: A better error message would be "Error generating state"

File: src/main/java/seedu/address/model/ModelManager.java [hunk 2]
```diff
@@ -121,12 +147,100 @@ public void setPerson(Person target, Person updatedPerson) {
...
+        }
+    }
```
Comment (Jolonauh) [inline src/main/java/seedu/address/model/ModelManager.java:216]: Add a newline below this line

File: src/main/java/seedu/address/model/ModelManager.java [hunk 3]
```diff
@@ -121,12 +147,100 @@ public void setPerson(Person target, Person updatedPerson) {
...
+     */
+    @Override
+    public void restoreState(State state) {
+        ObservableList<Person> newFilteredPersons = state.getFilteredList();
+        ReadOnlyAddressBook newAddressBook = state.getAddressBook();
+        Predicate<? super Person> newPredicate = state.getFilteredPersonsListPredicate();
+
+        setAddressBook(newAddressBook);
+        setFilteredPersonsList(newFilteredPersons, newPredicate);
+    }
```
Comment (Jolonauh) [inline src/main/java/seedu/address/model/ModelManager.java:228]: newline

File: src/test/java/seedu/address/history/HistoryManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    @Test
+    void addStateTest() {
+        history.addState(TYPICAL_SECOND_STATE);
+        assertEquals(history.getCurrState(), TYPICAL_SECOND_STATE);
+    }
+    @Test
+    void rollBackStateSuccess() {
+        history.addState(TYPICAL_SECOND_STATE);
+        assertDoesNotThrow(() -> history.rollBackState());
+    }
```
Comment (Jolonauh) [inline src/test/java/seedu/address/history/HistoryManagerTest.java:44]: Test names should follow the format : [what it does]_[input]_[expected outcome] 
e.g. write_emptyInput_exceptionThrown

**Review by Jolonauh — CHANGES_REQUESTED** [review]
Minor issues with formatting. Should also relook at the use of ReadOnlyAddressBook and ObservableList in the State class

File: src/main/java/seedu/address/history/State.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+
+/**
+ * The `State` class represents a snapshot of the HAL9000 application's state at a specific point in time.
+ * It contains information about the executed command and the list of tasks at that time.
+ */
+public class State {
+    private final Command command;
+    private final ReadOnlyAddressBook addressBook;
+    private final ObservableList<Person> filteredPersonsSource;
```
[the remark this answers is not in the document]
Comment (pjanthony2001) [reply src/main/java/seedu/address/history/State.java]: Fixed with most recent commit

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 6]
```diff
@@ -110,6 +110,7 @@ private void setAccelerator(MenuItem menuItem, KeyCombination keyCombination) {
      * Fills up all the placeholders of this window.
      */
     void fillInnerParts() {
+        System.out.println(logic.getFilteredPersonList());
```
Comment (Jolonauh) [inline src/main/java/seedu/address/ui/MainWindow.java:113]: Is this line for debugging? If so, it should be removed

**Review by Jolonauh — COMMENTED** [review]
LGTM

File: src/main/java/seedu/address/ui/MainWindow.java (hunk 6 above)
Comment (pjanthony2001) [reply src/main/java/seedu/address/ui/MainWindow.java:113]: oh yeah thanks mate I didn't catch that

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java [hunk 7]
```diff
@@ -82,6 +82,14 @@ public void toStringMethod() {
         assertEquals(expected, findCommand.toString());
     }
 
+    @Test
+    public void getCommandStringTest() {
+        NameContainsKeywordsPredicate predicate = preparePredicate("Kurz Elle Kunz"); //Dummy Predicate
```
Comment (tejas-1405) [inline src/test/java/seedu/address/logic/commands/FindCommandTest.java:87]: Can remove this if testing is done?

Comment (pjanthony2001) [reply src/test/java/seedu/address/logic/commands/FindCommandTest.java:87]: Done, removed the comment

**Pull request merged**